internal/controller/compute/statefulserverset/datavolume_controller.go

```diff
@@ -0,0 +1,165 @@
+	condemnedVolume, err := k.Get(ctx, name, namespace)
+	if err != nil {
+		return err
+	}
+	if err := k.kube.Delete(ctx, condemnedVolume); err != nil {
+		return fmt.Errorf("error deleting data volume %w", err)
+	}
+	return serverset.WaitForKubeResource(ctx, serverset.ResourceReadyTimeout, k.isDataVolumeDeleted, condemnedVolume.Name, namespace)
+}
+
+// isAvailable - checks if a volume is available
+func (k *kubeDataVolumeController) isAvailable(ctx context.Context, name, namespace string) (bool, error) {
+	obj, err := k.Get(ctx, name, namespace)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	if obj != nil && obj.Status.AtProvider.VolumeID != "" && strings.EqualFold(obj.Status.AtProvider.State, ionoscloud.Available) {
+		return true, nil
+	}
+	return false, err
+}
+
+func (k *kubeDataVolumeController) isDataVolumeDeleted(ctx context.Context, name, namespace string) (bool, error) {
+	k.log.Info("Checking if data volume is deleted", "name", name, "namespace", namespace)
+	obj := &v1alpha1.Volume{}
+	err := k.kube.Get(ctx, types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}, obj)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			k.log.Info("Data volume has been deleted", "name", name, "namespace", namespace)
+			return true, nil
+		}
+		return false, nil
+	}
+	return false, nil
+}
+
+func fromStatefulServerSetToVolume(cr *v1alpha1.StatefulServerSet, name string, replicaIndex, volumeIndex int) v1alpha1.Volume {
+	vol := v1alpha1.Volume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: cr.Namespace,
+			Labels: map[string]string{
+				statefulServerSetLabel: cr.Name,
+				// todo replace with function
+				fmt.Sprintf(volumeselector.VolumeReplicaIndexLabel, cr.Name+"-"+cr.Spec.ForProvider.Template.Metadata.Name, volumeselector.ResourceDataVolume): fmt.Sprintf("%d", replicaIndex),
+				fmt.Sprintf(volumeselector.VolumeIndexLabel, cr.Name+"-"+cr.Spec.ForProvider.Template.Metadata.Name, volumeselector.ResourceDataVolume):        fmt.Sprintf("%d", volumeIndex),
+			},
+		},
+		Spec: v1alpha1.VolumeSpec{
+			ResourceSpec: xpv1.ResourceSpec{
+				ProviderConfigReference: cr.GetProviderConfigReference(),
+				ManagementPolicies:      cr.GetManagementPolicies(),
+				DeletionPolicy:          cr.GetDeletionPolicy(),
+			},
+			ForProvider: v1alpha1.VolumeParameters{
+				DatacenterCfg:    cr.Spec.ForProvider.DatacenterCfg,
+				Name:             name,
+				AvailabilityZone: serverset.GetZoneFromIndex(replicaIndex),
+				Size:             cr.Spec.ForProvider.Volumes[volumeIndex].Spec.Size,
+				Type:             cr.Spec.ForProvider.Volumes[volumeIndex].Spec.Type,
+			},
+		}}
+	if cr.Spec.ForProvider.Volumes[volumeIndex].Spec.Image != "" {
+		vol.Spec.ForProvider.Image = cr.Spec.ForProvider.Volumes[volumeIndex].Spec.Image
+		vol.Spec.ForProvider.ImagePassword = "imagePassword776"
```

Comment:
Not sure if it is save to hardcode a password here

Reply:
will remove, images require passwords so it was for testing purposes. Same as for serverset